src/main/java/org/opensearch/knn/index/memory/NativeMemoryCacheManager.java

```diff
@@ -301,7 +306,47 @@ public NativeMemoryAllocation get(NativeMemoryEntryContext<?> nativeMemoryEntryC
             );
         }
 
-        return cache.get(nativeMemoryEntryContext.getKey(), nativeMemoryEntryContext::load);
+        if (KNNFeatureFlags.isForceEvictCacheEnabled()) {
+            // Utilizes a force eviction mechanism to free up memory before the entry can be added to the cache
+            // In case of a cache hit, the operation just updates the locally maintained recency list
+            // In case of a cache miss, least recently accessed entries are evicted in a blocking manner
+            // before the new entry can be added to the cache.
+            String key = nativeMemoryEntryContext.getKey();
+            NativeMemoryAllocation result = cache.getIfPresent(key);
+
+            // Cache Hit
+            if (result != null) {
+                accessRecencyQueue.remove(key);
```

Comment:
This is concerning part when there are many graphs in a single node. This remove is linear complexity and this code path will be hit for every query. I think the benchmark case might not have enough segment files to see the degradation.

Reply:
I think most of your concerns are around queue efficiency/operations. 

Let me see if I can replace it to a ordered set and make it work good enough. We do not have an out of the box concurrent ordered set solution.
We can always use `Collections.synchronizedSet(new LinkedHashSet<>())` - but the efficiency of wrapped concurrent collections is not as efficient.

Reply:
Can we consider utilizing removal listener?

Reply:
+1 on @heemin32 comment. Is there a way we can avoid having our own queue/set and use removal listener from Guava cache? Does Guava cache provides a sync way to remove the items from cache?

Reply:
That is the root of our problems - there are currently two async operations that happen for cache entry removal - 

1. The notification from Guava is not guaranteed to be before entry i.e the object cleanup can happen after an entry is added
2. The `close` on `NativeMemoryAllocation` objects is forked off into a separate thread (`onRemoval` from the cache is supposed to be sync with the removal flow)

This PR solves both of these problems. If we could utilize the listener directly - we would not be in this problem in the first place :) 

Details: https://github.com/google/guava/wiki/CachesExplained#when-does-cleanup-happen)

Reply:
I think removal listener should be called earlier enough unlike cleanup? Can we test it?
```
If your cache should not grow beyond a certain size, just use [CacheBuilder.maximumSize(long)](https://guava.dev/releases/snapshot/api/docs/com/google/common/cache/CacheBuilder.html#maximumSize-long-). The cache will try to evict entries that haven't been used recently or very often. Warning: the cache may evict entries before this limit is exceeded -- typically when the cache size is approaching the limit.
```

```
You may specify a removal listener for your cache to perform some operation when an entry is removed, via [CacheBuilder.removalListener(RemovalListener)](https://guava.dev/releases/snapshot/api/docs/com/google/common/cache/CacheBuilder.html#removalListener-com.google.common.cache.RemovalListener-). The [RemovalListener](https://guava.dev/releases/snapshot/api/docs/com/google/common/cache/RemovalListener.html) gets passed a [RemovalNotification](https://guava.dev/releases/snapshot/api/docs/com/google/common/cache/RemovalNotification.html), which specifies the [RemovalCause](https://guava.dev/releases/snapshot/api/docs/com/google/common/cache/RemovalCause.html), key, and value.
```

Reply:
Ran some more tests for the comparison between the two approaches (set vs optimized concurrent queue) - https://github.com/opensearch-project/k-NN/pull/2015#issuecomment-2327064680

It does look like the concurrent optimized version of the queue helps with perf.

Reply:
@heemin32 We do have an `onRemoval` listener attached: [Ref code](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/index/memory/NativeMemoryCacheManager.java#L383-L394)

The problem is that this is called after an entry is made into the cache - so there is still a potential for memory overflow - leading to node drops.

To validate this - I tested a theory using minimal changes from this solution
- Made no changes to the `get` method from `NativeMemoryCacheManager` 
- Made the `close()` method from `NativeMemoryAllocation` to be sync with the `onRemoval` method

This still led to memory overflows and node drop scenarios.
We need both the things to work in tandem.
The solutions are using an additional DS (like this solution) or design a cache that has strict enforcement.

Reply:
I would like to explore other cache option like caffeine or ehcache but I guess it can be discussed in a separate thread.

The main concern here is removing and adding entry in dequeue. 
Because this code path is called for every query, it could impact to query latency with high number of segments compared to current implementation.
One other approach would be having last access time in the cache item itself and updating it for every get operation. When eviction is needed, we scan the entire entry and evict least accessed item.

Reply:
The general idea here was to add in correctness.
A minor increase in latency is a lot better than node drops in my opinion. And this can be looked at as an incremental solution. 

With the feature flag turned off - you get the same performance as you do today. But for use cases where node drops are a bigger problem, we can turn on force eviction.

And your concerns are valid - this is not the most optimized solution currently. But it is a step in that direction.
We can work on improving it before enabling the feature by default.

Reply:
- Is there any urgency on this feature? Otherwise, implementing it without feature flag is the best.

Reply:
> Is there any urgency on this feature? Otherwise, implementing it without feature flag is the best.

Feature flag is there to be more of a defensive mechanism, since all tests cannot be covered and this code touches the hot code path. What I am not able to understand is why this question is coming up? Because this is a good feature that adds stability to cluster.

Reply:
> I would like to explore other cache option like caffeine or ehcache but I guess it can be discussed in a separate thread.

@heemin32 
Do you know that the limitations of Guava cache is not present in these caches?

Reply:
>Feature flag is there to be more of a defensive mechanism, since all tests cannot be covered and this code touches the hot code path. What I am not able to understand is why this question is coming up? Because this is a good feature that adds stability to cluster.

The question is more about the urgency not the need of this feature. I feel feature flag is mostly for user to configure the behavior. As a method of defensive mechanism, we should invest more on testing.
I think we can avoid using dequeue which puts latency for every query. Instead we can have last accessed time en cache entry and evict them when the cache is full where some latency won't have a big impact.

I haven't checked if they don't have the limitation of Guava cache.

Reply:
I see this as a bug in k-NN plugin, and with more. 


> As a method of defensive mechanism, we should invest more on testing.

There has been testing around it but not all cases can be covered. Hence having this helps.

> I think we can avoid using dequeue which puts latency for every query. Instead we can have last accessed time en cache entry and evict them when the cache is full where some latency won't have a big impact.

What you mean by this?


> I haven't checked if they don't have the limitation of Guava cache.

Then this is the best solution we have for now.